### DISCUSSION TITLE
Changes made in `OnAfterBootstrap` hook handler are not immediately applied

### DISCUSSION THREAD

**Discussion by bliepp** [pocketbase/pocketbase#4832] [Q&A] [answered]
Hey there,

I'm currently working on an app which I decided to write using pocketbase as framework. Here's my code:

```golang
package main
// ....
func main() {
	app := pocketbase.New()

        app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
		fmt.Println("Hello from here")
		return nil
	})
	app.OnAfterBootstrap().Add(AppSettingsHook)
	app.OnAfterBootstrap().Add(ModifyUsersCollectionHook)

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}

func AppSettingsHook(e *core.BootstrapEvent) error {
        if e.App.Settings().Meta.AppName != "Acme" {
		return nil
	}

	form := forms.NewSettingsUpsert(e.App)
	form.Meta.AppName = "My Custom App Name"
	form.Submit()

	return nil
}

func ModifyUsersCollectionHook(e *core.BootstrapEvent) error {
	users, err := e.App.Dao().FindCollectionByNameOrId("users")
	if err != nil {
		// users collection not found, ignore and proceed
		return nil
	}

	// rename avatar field to logo if not already done
	field := users.Schema.GetFieldByName("avatar")
	if field != nil {
		field.Name = "logo"
	}

	e.App.Dao().SaveCollection(users)

	return nil
}
```

I'm running it in a container.

So, what I'm trying to do is do a basic setup when starting the app (i.e. creating collections, filling in default data, setting the app name via code, etc.). However, there's a problem:

When I start in a freh environment (i.e. no `pb_data` directory) I'm prompted to create an admin user (as expected). After I created one and log in *none* of my changes are applied until I restart the container. Am I missing something? Further investigation shows that changes made in handlers of the `OnAfterBootstrap` hook are not applied immediately. However, The code is executed as proven with the first hook handler that only prints to stdout.

Any ideas?

bliepp

**Comment by ganigeorgiev** [comment, marked as answer]
It is not clear what do you mean by _"changes made in handlers of the OnAfterBootstrap hook are not applied immediately"_ but creating collections and initializing default settings should be done through [migrations](https://pocketbase.io/docs/go-migrations/) because the above code will execute every time when the app process start and this could cause side-effects like overwriting any previously saved setting changed from the UI.

Additionally note that currently the bootstrap hook executes **before** running the migrations (both system and app), meaning that your function `ModifyUsersCollectionHook` will return on the first condition since the "users" collection will not be created yet (it is created as part of the [system init migration](https://github.com/pocketbase/pocketbase/blob/master/migrations/1640988000_init.go#L103)).

_Keep in mind that the hooks execution order will slightly change with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355)._

**Reply by bliepp** [reply]
Thanks, that's very helpful information. Gonna dive further into it.

**Reply by bliepp** [reply]
Well, I feel a bit dumb for not recognizing migrations were exactly what I needed. Thank you!
